### ISSUE TITLE
InteropLowering and threading

### ISSUE THREAD

**Issue by kpgalligan** [JetBrains/kotlin-native#1601]
Not an issue. Feature question/discussion.

Out of the box, KN provides Worker as the primary artifact for multithreading. This will work in some general cases, but there are probably a number of scenarios which aren't easily covered. The most obvious (to me) is posting back to the main thread on iOS.

In general, once you get down to the C++ level there's a lot of flexibility. One major exception is passing a function across threads to implement a different threading handler. The modification for Worker on schedule is pretty specific to its purpose. I've tried staticCFunction, but it has restrictions on types that Worker does not.

To be clear, I assume staticCFunction restricts types because it needs to function in various interop situations, but I'm talking about having Kotlin talk to C++ in a manner similar to how runtime does, so I assume whatever issue is involved with general interop doesn't apply.

I'm still fairly new to the C++ area, so there may be a way to do this without hacking the compiler, but it looks like the InteropLowering checks are hard coded to types available in the runtime module.

## Option 1
Provide some kind of hook into the compiler to allow extensions to lowering. I'm definitely not an expert on that area of the code, so I assume that might be difficult.

## Option 2
Adding extensions to runtime and the compiler for specific cases.

I added a method and modified the compiler to try out option 2. The method looks like this:

```kotlin
fun <T1, T2> threadSafe(@VolatileLambda job: (T1) -> T2): CPointer<CFunction<(T1) -> T2>> =...
```

The code usage looks something like this:

```kotlin
//Called from main thread
dispatchSingle<SomeData, SomeData>(
                {SomeData(21)},

                //Runs in another thread
                threadSafe {
                    initRuntimeIfNeeded()
                    SomeData(it.a * 2)
                },

                //Runs in main thread
                {
                    println("The answer ${it.a}")
                }
        )
```

The modification to InteropTransformer looks like the code for staticCFunction without the checks:

```kotlin
in interop.threadSafeFunction -> {
                val irCallableReference = unwrapStaticFunctionArgument(expression.getValueArgument(0)!!)

                if (irCallableReference == null || irCallableReference.getArguments().isNotEmpty()) {
                    context.reportCompilationError(
                            "${descriptor.fqNameSafe} must take an unbound, non-capturing function or lambda",
                            irFile, expression
                    )
                }

                val targetSymbol = irCallableReference.symbol
                val target = targetSymbol.descriptor

                IrFunctionReferenceImpl(
                        builder.startOffset, builder.endOffset,
                        expression.type,
                        targetSymbol, target,
                        typeArguments = null)
            }
```

This works for my simple test case, although I haven't done anything more sophisticated with it. Thoughts? Alternative ways to do this?

**Comment by SvyatoslavScherbina** [comment]
Actually, we are designing some solutions and approaches for the wide variety of situations like this.
This case is not very complicated, and already have the temporary solution: instead of making pointer to arbitrary Kotlin function you can create a usual lambda and then freeze it. It should be mentioned that all values captured by such a lambda would get frozen too, but it is OK if lambda captures only frozen (or otherwise shareable) objects.
Also we have `detachObjectGraph`/`attachObjectGraph` APIs allowing to decouple object transfer from job scheduling: https://github.com/JetBrains/kotlin-native/blob/54fa7468138eb6b76ac8686e6bc3718f045645bd/runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt#L65

To summarize all of the above, let me transform your code snippet a bit:
```
//Called from main thread
val data = detachObjectGraph { SomeData() }

dispatchToBg({ //Runs in another thread
    val it = attachObjectGraph<SomeData>(data)
    val result  = detachObjectGraph { SomeData(it.a * 2) }
    dispatchToMain({ //Runs in main thread
        val it = attachObjectGraph<SomeData>(result)
        showText("The answer ${it.a}")
    }.freeze())
}.freeze()),
```

Of course it can be encapsulated to provide the API from your original snippet. What I'd like to emphasize that it is already possible to achieve the same goal by writing code in Kotlin without introducing any new intrinsics.

**Comment by kpgalligan** [comment]
Thanks. I'll play around with this and see how it goes. Either freezing or detaching should work for now.
